streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -656,6 +657,71 @@ boolean tryToCompleteRestoration(final long now, final java.util.function.Consum
         return allRunning;
     }
 
+    private void addTaskstoStateUpdater() {
+        for (final Task task : tasks.drainPendingTaskToInit()) {
+            task.initializeIfNeeded();
+            stateUpdater.add(task);
+        }
+    }
+
+    private void handleRemovedTasksFromStateUpdater() {
+        final Map<TaskId, RuntimeException> taskExceptions = new LinkedHashMap<>();
+        final Set<Task> tasksToCloseDirty = new TreeSet<>(Comparator.comparing(Task::id));
+
+        for (final Task task : stateUpdater.drainRemovedTasks()) {
+            final TaskId taskId = task.id();
+            Set<TopicPartition> inputPartitions;
+            if ((inputPartitions = tasks.removePendingTaskToRecycle(task.id())) != null) {
+                try {
+                    final Task newTask = task.isActive() ?
+                        convertActiveToStandby((StreamTask) task, inputPartitions) :
+                        convertStandbyToActive((StandbyTask) task, inputPartitions);
+                    newTask.initializeIfNeeded();
+                    stateUpdater.add(newTask);
+                } catch (final RuntimeException e) {
+                    final String uncleanMessage = String.format("Failed to recycle task %s cleanly. " +
+                        "Attempting to close remaining tasks before re-throwing:", taskId);
+                    log.error(uncleanMessage, e);
+
+                    if (task.state() != State.CLOSED) {
+                        tasksToCloseDirty.add(task);
+                    }
+
+                    taskExceptions.putIfAbsent(taskId, e);
+                }
+            } else if (tasks.removePendingTaskToClose(task.id())) {
+                try {
+                    task.closeClean();
+                    if (task.isActive()) {
+                        activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(task.id());
+                    }
+                } catch (final RuntimeException e) {
+                    final String uncleanMessage = String.format("Failed to close task %s cleanly. " +
+                        "Attempting to close remaining tasks before re-throwing:", task.id());
+                    log.error(uncleanMessage, e);
+
+                    if (task.state() != State.CLOSED) {
+                        tasksToCloseDirty.add(task);
+                    }
+
+                    taskExceptions.putIfAbsent(task.id(), e);
+                }
+            } else if ((inputPartitions = tasks.removePendingTaskToUpdateInputPartitions(task.id())) != null) {
+                task.updateInputPartitions(inputPartitions, topologyMetadata.nodeToSourceTopics(task.id()));
+                stateUpdater.add(task);
```

Comment:
Found this issue when adding unit test: after update the input partitions, need to add it back to state updater.